rangeserver/src/cache/memtabledb.rs

```diff
@@ -220,17 +217,24 @@ impl Cache for MemTableDB {
             },
         }
 
-        for i in 0..self.rc_options.num_write_buffers {
-            let id = (self.mutable_id + self.rc_options.num_write_buffers - i) % self.rc_options.num_write_buffers; // mut_id is usize, should not become negative
-            // let memtable_rg = self.memtables[id].read().unwrap();
-            // let memtable_rg = self.memtables[id].read().await;
-            println!("GET-DBG: id {}, epoch {}, oldest_epoch {}, latest_epoch {}", id, epoch_internal, 
-                self.memtables[id].oldest_epoch, self.memtables[id].latest_epoch);
-            if self.memtables[id].status == MemTableStatus::FLUSHED {
+        for i in 0..self.cache_options.num_write_buffers {
+            let idx = (self.mut_memtable_idx + self.cache_options.num_write_buffers - i) % self.cache_options.num_write_buffers; // mut_id is usize, should not become negative
+            println!("GET-DBG: key {:?} memtable_idx {}, epoch {}, oldest_epoch {}, latest_epoch {}", key, idx, epoch_internal, self.memtables[idx].oldest_epoch, self.memtables[idx].latest_epoch);
+            if self.memtables[idx].status == MemTableStatus::FLUSHED {
                 break
             }
-            if self.memtables[id].oldest_epoch <= epoch_internal && epoch_internal <= self.memtables[id].latest_epoch {                
-                match self.memtables[id].entries.iter().find(|&&ref entry| entry.key == key && entry.epoch <= epoch_internal)  {
+            if self.memtables[idx].oldest_epoch <= epoch_internal && epoch_internal <= self.memtables[idx].latest_epoch {
+                let start = MemTableEntry{
+                    key: key.clone(),
+                    val: Vec::new().into(),
+                    epoch: epoch_internal,
+                    deleted: false};
+                let mut end = start.clone();
+                end.epoch = 0;
+
+                let result = self.memtables[idx].entries.range(Included(&start), Included(&end)).next();
```

Comment:
This is equivalent to calling `lower_bound()` I think

Reply:
If I change the code to: `let result = self.memtables[idx].entries.lower_bound(Included(&start));`

This test case does not work.
- memtable with oldest_epoch 3284, latest_epoch 3299. There is a one to one association with key to epoch i.e. k1 is epoch 1, k2 is epoch2 and so on.
- get(key: "k0"  epoch: 3299) or get(key: "k0"  epoch: None) 
The key should not be found since the smallest key in this memtable is k3284 for epoch 3284.

But, lower_bound instead of returning None, returns the smallest entry MemTableEntry { key: b"k3284", val: b"\xd4\0\0\0\0\0\0\0\0\0", epoch: 3284, deleted: false }. I will have to do another check on the returned key to make sure the key matches.

Maybe I am missing something?

Reply:
That's expected. 
```
Returns an Option<&T> pointing to the lowest element whose key is above the given bound. If no such element is found then None is returned.
```

In this case, in that memtable the lowest element whose key is above `k0` is `k3284`. So that is returned. If all keys were smaller than `k0` then `None` would be returned. Think of `lower_bound()` as asking for keys `>=` the parameter. You do have to perform a comparison afterwards. I think it might still be better than `range()` because there would be no need to compare with the `max` parameter.

Reply:
switched to lower_bound with key check